icu4c/source/common/rbbi_cache.cpp

```diff
@@ -420,50 +426,141 @@ UBool RuleBasedBreakIterator::BreakCache::populateNear(int32_t position, UErrorC
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet numberSet;
+    if(numberSet.isEmpty()) {
+        numberSet.applyPattern(UnicodeString(u"[:Nd:]"), status);
+        if (U_SUCCESS(status)) {
+            numberSet.compact();
+        }
+    }
+    return numberSet.contains(ch);
+}
+
+static inline bool isCj(UChar32 ch) {
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet cjSet;
+    if(cjSet.isEmpty()) {
+        cjSet.applyPattern(UnicodeString(u"[[:Han:][:Hiragana:][:Katakana:]\u30fc\uff70\uff9e\uff9f]"), status);
+        if (U_SUCCESS(status)) {
+            cjSet.compact();
+        }
+    }
+    return cjSet.contains(ch);
+}
+
+/* getting a uint32_t properties word from the data */
+#define GET_PROPS(c, result) ((result)=UTRIE2_GET16(&propsTrie, c))
+
+/** Mask constant for multiple UCharCategory bits (OP Open Punctuation). @stable ICU 2.1 */
+#define U_GC_OP_MASK \
+            (U_GC_PI_MASK|U_GC_PS_MASK)
+#define U_GC_CP_MASK \
+            (U_GC_PD_MASK|U_GC_PE_MASK|U_GC_PC_MASK|U_GC_PO_MASK|U_GC_PF_MASK)
+
+static UBool u_isOpenPunct(UChar32 c) {
+    uint32_t props;
+    GET_PROPS(c, props);
+    return (UBool)((CAT_MASK(props)&U_GC_OP_MASK)!=0);
+}
+
+static UBool u_isClosePunct(UChar32 c) {
+    uint32_t props;
+    GET_PROPS(c, props);
+    return (UBool)((CAT_MASK(props)&U_GC_CP_MASK)!=0);
+}
+
+UBool RuleBasedBreakIterator::BreakCache::handleNumber(int32_t pos) {
+    return pos < utext_nativeLength(&fBI->fText)
+            && pos > 0
+            && isNumber(utext_char32At(&fBI->fText, pos-1))
+            && (isCj(utext_char32At(&fBI->fText, pos))
+            || u_isClosePunct(utext_char32At(&fBI->fText, pos)));
+}
 
+UBool RuleBasedBreakIterator::BreakCache::handleClosePunctuation(int32_t pos) {
+    return ( pos < utext_nativeLength(&fBI->fText)
+            && pos > 0
+            && u_isClosePunct(utext_char32At(&fBI->fText, pos))
+            && (isCj(utext_char32At(&fBI->fText, pos-1))
+            || u_isClosePunct(utext_char32At(&fBI->fText, pos-1))));
+}
 
 UBool RuleBasedBreakIterator::BreakCache::populateFollowing() {
     int32_t fromPosition = fBoundaries[fEndBufIdx];
     int32_t fromRuleStatusIdx = fStatuses[fEndBufIdx];
     int32_t pos = 0;
     int32_t ruleStatusIdx = 0;
 
-    if (fBI->fDictionaryCache->following(fromPosition, &pos, &ruleStatusIdx)) {
-        addFollowing(pos, ruleStatusIdx, UpdateCachePosition);
-        return TRUE;
+    while (fBI->fDictionaryCache->following(fromPosition, &pos, &ruleStatusIdx)) {
+        if(!fBI->fIsPhraseBreaking || (!handleClosePunctuation(pos) && !handleNumber(pos))) {
+            addFollowing(pos, ruleStatusIdx, UpdateCachePosition);
+            return TRUE;
+        } else {
+            fromPosition = pos;
+        }
     }
 
     fBI->fPosition = fromPosition;
-    pos = fBI->handleNext();
-    if (pos == UBRK_DONE) {
-        return FALSE;
-    }
+    do {
+        pos = fBI->handleNext();
+        while(fBI->fIsPhraseBreaking
```

Comment:
I suspect that the situation being addressed by this (and the following) logic may be a somewhat more wide-spread problem with any dictionary-based line breaking. The line breaking rules regarding open & close punctuation all fall apart once a dictionary is involved. Again, needs some further thought.